workload/schemachange: fix date parsing error in SET DEFAULT operations

spilchen opened the pull request:
The schemachange workload could generate invalid SQL when testing type mismatches, causing parse errors instead of expected type mismatch errors.

Fix by always using explicit type casting in SET DEFAULT statements.

Fixes: #152572

Release note: None

Epic: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/152877)
<!-- Reviewable:end -->

rafiss reviewed (approved):
nice find!

spilchen committed:
workload/schemachange: fix date parsing error in SET DEFAULT operations

The schemachange workload could generate invalid SQL when testing type
mismatches, causing parse errors instead of expected type mismatch errors.

Fix by always using explicit type casting in SET DEFAULT statements.

Fixes: #152572

Release note: None

Epic: None

spilchen commented on pkg/workload/schemachange/operation_generator.go:
```diff
@@ -2530,8 +2530,11 @@ func (og *operationGenerator) setColumnDefault(ctx context.Context, tx pgx.Tx) (
+	// Always use explicit type casting to ensure consistent behavior and avoid parse errors.
+	// When the types don't match, this will produce the expected DatatypeMismatch error.
+	// When the types do match, the cast is harmless and makes the intent explicit.
+	stmt.sql = fmt.Sprintf(`ALTER TABLE %s ALTER COLUMN %s SET DEFAULT %s:::%s`,
```
Good to know. I had seen both and didn't realize the difference. I updated it to be a cast.
<!-- Reviewable comment -OZELPQ90nNElGN5gm8Z:b-gj4jns -->
<!-- Sent from Reviewable.io -->

spilchen reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/152877)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @herkolategan, @rafiss, and @srosenberg)



<!-- Sent from Reviewable.io -->
